File: hash/coprime_products/coprime_products.cpp
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;

class CoprimeProducts
{ // LeetCode Q.2584.
private:
    unordered_map<int, int> findPrimes(int num)
    {
        unordered_map<int, int> primes2Counts; // Counts of all occurred primes.
        while (num % 2 == 0)
        {
            primes2Counts[2]++;
            num /= 2;
        }

        // Num must be odd now. Increment skips even elements.
        for (int factor = 3; factor * factor <= num; factor += 2)
        {
            while (num % factor == 0)
            {
                primes2Counts[factor]++;
                num /= factor;
            }
        }

        if (num > 2) // When num is a prime greater than 2.
            primes2Counts[num]++;

        return primes2Counts;
    }

public:
    int find_split_idx(vector<int> &nums)
    {
        unordered_map<int, unordered_map<int, int>> nums2Primes;
        unordered_map<int, int> prefixPrimes, suffixPrimes;

        for (auto num : nums)
        {
            if (nums2Primes.find(num) == nums2Primes.end())
                nums2Primes[num] = findPrimes(num);

            for (auto &pair : nums2Primes[num])
            {
                auto [prime, count] = pair;
                suffixPrimes[prime] += count;
            }
        }

        unordered_set<int> commonPrimes;
        for (int idx = 0; idx < nums.size() - 1; idx++)
        {
            for (auto &pair : nums2Primes[nums[idx]])
            {
                auto [prime, count] = pair;
                prefixPrimes[prime] += count;
                commonPrimes.insert(prime);
            }

            for (auto &pair : nums2Primes[nums[idx]])
            {
                auto [prime, count] = pair;
                suffixPrimes[prime] -= count;
                if (suffixPrimes[prime] == 0)
                {
                    if (commonPrimes.find(prime) != commonPrimes.end())
                        commonPrimes.erase(prime);
                }
            }

            if (commonPrimes.empty()) // Current idx is the smallest idx to split.
                return idx;
        }
        return -1;
    }
};

File: hash/coprime_products/coprime_products.py

def _find_primes(num: int) -> dict[int, int]:
    primes2counts: dict[int, int] = dict()  # Counts of all occurred primes.
    while num % 2 == 0:
        if 2 not in primes2counts.keys():
            primes2counts.update({2: 0})
        primes2counts[2] += 1
        num /= 2

    factor = 3  # Num must be odd now. Increment skips even elements.
    while factor * factor <= num:
        while num % factor == 0:
            if factor not in primes2counts.keys():
                primes2counts.update({factor: 0})
            primes2counts[factor] += 1
            num /= factor

        factor += 2

    if num > 2:  # When num is a prime greater than 2.
        primes2counts[num] = 1

    return primes2counts


def find_split_idx(nums: list[int]) -> int:  # LeetCode Q.2584.
    nums2primes: dict[int, dict[int, int]] = dict()
    prefix_primes: dict[int, int] = dict()
    suffix_primes: dict[int, int] = dict()

    for num in nums:
        if num not in nums2primes.keys():
            nums2primes[num] = _find_primes(num)

        for prime, count in nums2primes[num].items():
            if prime not in suffix_primes.keys():
                suffix_primes.update({prime: 0})
            suffix_primes[prime] += count

    common_primes: set[int] = set()
    for idx in range(len(nums) - 1):
        for prime, count in nums2primes[nums[idx]].items():
            if prime not in prefix_primes.keys():
                prefix_primes.update({prime: 0})
            prefix_primes[prime] += count
            common_primes.add(prime)

        for prime, count in nums2primes[nums[idx]].items():
            suffix_primes[prime] -= count
            if suffix_primes[prime] == 0:
                if prime in common_primes:
                    common_primes.remove(prime)

        if not common_primes:  # Current idx is the smallest idx to split.
            return idx

    return -1
